ocaml/typing/typedecl.ml

```diff
@@ -1018,6 +1019,15 @@ let rec check_constraints_rec env loc visited ty =
   | Tpoly (ty, tl) ->
       let _, ty = Ctype.instance_poly false tl ty in
       check_constraints_rec env loc visited ty
+  | Tfunctor (_, id, (p, fl), ty) ->
+      List.iter (fun (_, t) -> check_constraints_rec env loc visited t) fl;
+      let mty = !Ctype.modtype_of_package env loc p fl in
+      let id' = Ident.(create_scoped ~scope:lowest_scope (name_unscoped id)) in
+      let env = Env.add_module id' Mp_present mty env in
+      let ty = Option.value ~default:ty
+          (Ctype.instance_funct ~id_in:(Ident.of_unscoped id)
+              ~p_out:(Pident id') ~fixed:false ty) in
+      check_constraints_rec env loc visited ty
```

Comment:
We need to have an instantiation here in order to do unification later on when checking the constraints (we unify a fresh unification variable with the relevant type). The instantiation also happen with `Tpoly` in the lines just above

Reply:
> Also, this function uses a visited set for recursion, whilst this code will create fresh type nodes. Are we confident this will always terminate? Are there tests for the case that might struggle to terminate?

The instantiation only goes directly to the use cases of the correct identifier. Thus it won't go recursively inside the same binding once again and refresh itself